dptb/postprocess/bandstructure/band.py

```diff
@@ -239,48 +203,49 @@ def get_bands(self, data: Union[AtomicData, ase.Atoms, str], kpath_kwargs: dict,
             high_sym_kpoints = kpath_kwargs.get('high_sym_kpoints', None)
             xlist = kpath_kwargs.get('xlist', None)
             labels = kpath_kwargs.get('labels', None)
+
         else:
             log.error('Error, now, kline_type only support ase_kpath, abacus, or vasp.')
             raise ValueError
         
-        # set the kpoint of the AtomicData
-        data[AtomicDataDict.KPOINT_KEY] = torch.nested.as_nested_tensor([torch.as_tensor(klist, dtype=self.model.dtype, device=self.device)])
-
-        # get the eigenvalues
-        data = self.model(data)
-        if self.overlap == True:
-            assert data.get(AtomicDataDict.EDGE_OVERLAP_KEY) is not None
-        data = self.eigv(data)
+        _, eigenvalues = self.get_eigs(data, klist, AtomicData_options)
+        
 
         # get the E_fermi from data
         nel_atom = kpath_kwargs.get('nel_atom', None)
-        assert isinstance(nel_atom, dict) or nel_atom is None
+        # assert isinstance(nel_atom, dict) or nel_atom is None
         
+        # if nel_atom is not None:
+        #     atomtype_list = self.data[AtomicDataDict.ATOM_TYPE_KEY].flatten().tolist()
+        #     atomtype_symbols = np.asarray(self.model.idp.type_names)[atomtype_list].tolist()
+        #     total_nel = np.array([nel_atom[s] for s in atomtype_symbols]).sum()
+        #     if hasattr(self.model,'soc_param'):
+        #         spindeg = 1
+        #     else:
+        #         spindeg = 2
+        #     estimated_E_fermi = self.estimate_E_fermi(self.data[AtomicDataDict.ENERGY_EIGENVALUE_KEY][0].detach().cpu().numpy(), total_nel, spindeg)
+        #     log.info(f'Estimated E_fermi: {estimated_E_fermi} based on the valence electrons setting nel_atom : {nel_atom} .')
+        # else:
+        #     estimated_E_fermi = None
         if nel_atom is not None:
-            atomtype_list = data[AtomicDataDict.ATOM_TYPE_KEY].flatten().tolist()
-            atomtype_symbols = np.asarray(self.model.idp.type_names)[atomtype_list].tolist()
-            total_nel = np.array([nel_atom[s] for s in atomtype_symbols]).sum()
-            if hasattr(self.model,'soc_param'):
-                spindeg = 1
-            else:
-                spindeg = 2
-            estimated_E_fermi = self.estimate_E_fermi(data[AtomicDataDict.ENERGY_EIGENVALUE_KEY][0].detach().cpu().numpy(), total_nel, spindeg)
-            log.info(f'Estimated E_fermi: {estimated_E_fermi} based on the valence electrons setting nel_atom : {nel_atom} .')
+            _,estimated_E_fermi = self.get_fermi_level(data=structase, nel_atom=nel_atom, \
```

Comment:
The eigenvalue seems computed twice here. In the second time, it should be possible to pass the eigenvalues from the first time and estimate without diagonalization again?